Provide wrappers for some ObjC Runtime functions that return a MallocSpan

cdumez opened the pull request:
#### a2691968c2351473b254a2cda09371433ddd4928
<pre>
Provide wrappers for some ObjC Runtime functions that return a MallocSpan
<a href="https://bugs.webkit.org/show_bug.cgi?id=282480">https://bugs.webkit.org/show_bug.cgi?id=282480</a>

Reviewed by Darin Adler.

* Source/JavaScriptCore/API/ObjcRuntimeExtras.h:
(protocolImplementsProtocol):
(forEachProtocolImplementingProtocol):
(forEachMethodInClass):
(forEachMethodInProtocol):
(forEachPropertyInProtocol):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/ObjCRuntimeExtras.h:
(WTF::class_copyMethodListSpan):
(WTF::class_copyProtocolListSpan):
(WTF::protocol_copyMethodDescriptionListSpan):
(WTF::protocol_copyPropertyListSpan):
(WTF::protocol_copyProtocolListSpan):
* Source/WebCore/bridge/objc/objc_class.mm:
(JSC::Bindings::ObjcClass::methodNamed const):
* Source/WebCore/platform/ios/WebAVPlayerController.mm:
(createWebAVPlayerControllerForwarderClass):
* Source/WebKit/Shared/API/Cocoa/_WKRemoteObjectInterface.mm:
(initializeMethods):

Canonical link: <a href="https://commits.webkit.org/286064@main">https://commits.webkit.org/286064@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de969a910f4317e922b468037243979b821383d9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74726 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54156 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27543 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79148 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25964 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76843 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63292 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1941 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/58698 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16982 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77793 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48846 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64217 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39092 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46065 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/21714 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24297 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/67863 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67281 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22059 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80639 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/73984 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2044 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1210 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/66957 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2192 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64235 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66248 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10192 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8344 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/96255 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11530 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2009 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4796 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21039 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2037 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/2957 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2044 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/API/ObjcRuntimeExtras.h:
```diff
@@ -60,9 +60,8 @@ inline void forEachProtocolImplementingProtocol(Class cls, Protocol *target, voi
-        auto protocols = adoptSystem<__unsafe_unretained Protocol*[]>(class_copyProtocolList(cls, &protocolsCount));
-        worklist.append(std::span { protocols.get(), protocolsCount });
+        auto protocols = class_copyProtocolListSpan(cls);
+        worklist.append(protocols.span());
```
Better as a single line with no braces:
```
worklist.append(class_copyProtocolListSpan(cls).span());
```

darinadler commented on Source/JavaScriptCore/API/ObjcRuntimeExtras.h:
```diff
@@ -83,35 +82,31 @@ inline void forEachProtocolImplementingProtocol(Class cls, Protocol *target, voi
-            auto protocols = adoptSystem<__unsafe_unretained Protocol*[]>(protocol_copyProtocolList(protocol, &protocolsCount));
-            worklist.append(std::span { protocols.get(), protocolsCount });
+            auto protocols = protocol_copyProtocolListSpan(protocol);
+            worklist.append(protocols.span());
```
Better as a single line with no braces:
```
worklist.append(class_copyProtocolListSpan(protocols).span());
```

darinadler commented on Source/WTF/wtf/ObjCRuntimeExtras.h:
```diff
@@ -41,3 +45,48 @@ inline ReturnType wtfCallIMP(IMP implementation, id target, SEL selector, Argume
 
 #endif // __cplusplus
+
+namespace WTF {
```
Seems like this needs to be inside the `#if __cplusplus` conditional.

darinadler commented on Source/WTF/wtf/ObjCRuntimeExtras.h:
```diff
@@ -41,3 +45,48 @@ inline ReturnType wtfCallIMP(IMP implementation, id target, SEL selector, Argume
+
+namespace WTF {
+
+inline MallocSpan<Method, SystemMalloc> class_copyMethodListSpan(Class cls)
```
I understand that you are doing inlines here to avoid overhead and perhaps avoid a `.cpp` file. I’d be tempted to not do that because I don’t think function call overhead is significant compared to `malloc`/`free` and I try to avoid putting our code into headers unless it’s an important optimization.

darinadler reviewed (approved):
Nice improvement; love it

cdumez committed:
Provide wrappers for some ObjC Runtime functions that return a MallocSpan
https://bugs.webkit.org/show_bug.cgi?id=282480

Reviewed by Darin Adler.

* Source/JavaScriptCore/API/ObjcRuntimeExtras.h:
(protocolImplementsProtocol):
(forEachProtocolImplementingProtocol):
(forEachMethodInClass):
(forEachMethodInProtocol):
(forEachPropertyInProtocol):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/ObjCRuntimeExtras.h:
(WTF::class_copyMethodListSpan):
(WTF::class_copyProtocolListSpan):
(WTF::protocol_copyMethodDescriptionListSpan):
(WTF::protocol_copyPropertyListSpan):
(WTF::protocol_copyProtocolListSpan):
* Source/WebCore/bridge/objc/objc_class.mm:
(JSC::Bindings::ObjcClass::methodNamed const):
* Source/WebCore/platform/ios/WebAVPlayerController.mm:
(createWebAVPlayerControllerForwarderClass):
* Source/WebKit/Shared/API/Cocoa/_WKRemoteObjectInterface.mm:
(initializeMethods):

Canonical link: https://commits.webkit.org/286064@main
